### PR TITLE
Form schema

### PR DESCRIPTION
**Pull request by chillu**
Implements Form Schema RFC: https://github.com/silverstripe/silverstripe-framework/issues/4938.
Relies on https://github.com/silverstripe/silverstripe-cms/pull/1410

The CMS integration isn't actually used yet, since no client-side code is sending the required `X-Formschema-Request` header. This will soon be the case once we make use of the schema in `AssetAdmin`, still in feature branches:
https://github.com/open-sausages/silverstripe-framework/tree/integrate-react
https://github.com/open-sausages/silverstripe-cms/tree/integrate-react
https://github.com/open-sausages/silverstripe-asset-admin

The easiest way to test its operation for now is using `curl`:
- Perform a CMS action
- Right click on the ajax request in the "network" dev tools tab, "copy as curl"
- Paste in terminal, add the following: `-H 'X-Formschema-Request: schema,state'`

Tests to perform:
- Save a folder in admin/assets
- `GET admin/assets/schema/EditForm`
- Save a folder with a validation error in admin/assets. This requires the following addition to `Folder.php`:

```
public function getCMSValidator() {
	return new RequiredFields('Name');
}
```

I'm not super happy with the solution:
- Too much copypaste with `setValidationResponseCallback()`
- Adds more `FormField` public API surface. Discussed with @sminnee and @tractorcow and decided to at least mark the API as experimental. 
- Messy CMS response negotiation. We have to make this work in parallel to PJAX responses for now, since only admin/assets will be React powered in 4.x. 

Long term, I want to refactor `Form` and `FormField` to become pure value objects, with form handling and validation logic being moved to the controller. Presentation into different formats (HTML, JSON, etc) could then also be handled in the controller layer, or in a component which is decoupled from `Form`. This would make `FormSchema` use a lot cleaner in regards to CMS controllers. But it requires changes as deep as `DataObject->getCMSFields()`, so wasn't really feasible in the context of this PR.

### CODE REVIEW

File: admin/code/LeftAndMain.php [hunk 1]
```diff
@@ -1260,14 +1353,27 @@ public function getEditForm($id = null, $fields = null) {
...
+			$negotiator = $this->getResponseNegotiator();
+			$form = Form::create(
 				$this, "EditForm", $fields, $actions
 			)->setHTMLID('Form_EditForm');
-			$form->setResponseNegotiator($this->getResponseNegotiator());
 			$form->addExtraClass('cms-edit-form');
 			$form->loadDataFrom($record);
 			$form->setTemplate($this->getTemplatesWithSuffix('_EditForm'));
 			$form->setAttribute('data-pjax-fragment', 'CurrentForm');
+			$form->setValidationResponseCallback(function() use ($negotiator, $form) {
```
Comment (tractorcow) [inline admin/code/LeftAndMain.php]: Won't you need to set this callback on the other leftandmain subclasses too? E.g. SecurityAdmin swaps out CMSForm for Form, but doesn't have this code. Also ModelAdmin.

File: forms/FormField.php [hunk 2]
```diff
@@ -162,7 +164,30 @@ class FormField extends RequestHandler {
...
 	 */
-	protected $attributes = array();
+	protected $attributes = [];
+
+	/**
+	 * The type of front-end component to render the FormField as.
+	 *
+	 * @var string
+	 */
+	protected $schemaComponent;
```
Comment (tractorcow) [inline forms/FormField.php]: Why aren't all these $schema* properties in the trait?

**Comment by tractorcow** [comment]
Tests stalled; I've restarted them.

File: admin/code/LeftAndMain.php (hunk 1 above)
Comment (chillu) [inline admin/code/LeftAndMain.php]: `SecurityAdmin` uses `parent::getEditForm()` which gets the callback from `LeftAndMain`.
`ModelAdmin` uses `GridFieldDetailForm`, which in turn still uses its own PJAX logic (in `generateValidationResponse()`), I've confirmed a ModelAdmin can still display validation errors.
Once these interfaces are switched over to React, requiring JSON responses, we can look into migrating the new logic there.

File: forms/FormField.php (hunk 2 above)
Comment (chillu) [inline forms/FormField.php]: Fixed now

**Comment by tractorcow** [comment]
@chillu still need that branch?

**Comment by chillu** [comment]
Nope, I've removed the branch. David will rebase on master for his JS-based form schema work

**Pull request merged**